p11-kit/add-profile.c

```diff
@@ -0,0 +1,240 @@
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * Author: Zoltan Fridrich <zfridric@redhat.com>
+ */
+
+#include "config.h"
+
+#include "constants.h"
+#include "debug.h"
+#include "iter.h"
+#include "message.h"
+#include "tool.h"
+
+#include <assert.h>
+#include <stdlib.h>
+
+#ifdef ENABLE_NLS
+#include <libintl.h>
+#define _(x) dgettext(PACKAGE_NAME, x)
+#else
+#define _(x) (x)
+#endif
+
+int
+p11_kit_add_profile (int argc,
+		     char *argv[]);
+
+static int
+add_profile (const char *token_str,
+	     CK_PROFILE_ID profile)
+{
+	int ret = 1;
+	CK_RV rv;
+	CK_OBJECT_HANDLE object = 0;
+	CK_SESSION_HANDLE session = 0;
+	CK_FUNCTION_LIST *module = NULL;
+	CK_FUNCTION_LIST **modules = NULL;
+	P11KitUri *uri = NULL;
+	P11KitIter *iter = NULL;
+	CK_OBJECT_CLASS klass = CKO_PROFILE;
+	CK_PROFILE_ID profile_id = CKP_INVALID_ID;
+	CK_ATTRIBUTE matching = { CKA_CLASS, &klass, sizeof (klass) };
+	CK_ATTRIBUTE attr = { CKA_PROFILE_ID, &profile_id, sizeof (profile_id) };
+	CK_ATTRIBUTE object_template[] = {
+	    { CKA_CLASS, &klass, sizeof (klass) },
+	    { CKA_PROFILE_ID, &profile, sizeof (profile) }
+	};
+
+	uri = p11_kit_uri_new ();
+	if (uri == NULL) {
+		p11_message (_("failed to allocate memory for URI"));
+		goto cleanup;
+	}
+
+	if (p11_kit_uri_parse (token_str, P11_KIT_URI_FOR_TOKEN, uri) != P11_KIT_URI_OK) {
+		p11_message (_("failed to parse the token URI"));
+		goto cleanup;
+	}
+
+	modules = p11_kit_modules_load_and_initialize (0);
+	if (modules == NULL) {
+		p11_message (_("failed to load and initialize modules"));
+		goto cleanup;
+	}
+
+	iter = p11_kit_iter_new (uri, 0);
+	if (iter == NULL) {
+		p11_message (_("failed to initialize iterator"));
+		goto cleanup;
+	}
+
+	p11_kit_iter_add_filter (iter, &matching, 1);
+	p11_kit_iter_begin (iter, modules);
+	while ((rv = p11_kit_iter_next (iter)) == CKR_OK) {
+		rv = p11_kit_iter_get_attributes (iter, &attr, 1);
+		if (rv != CKR_OK) {
+			p11_message (_("failed to retrieve attribute of an object"));
+			goto cleanup;
+		}
+
+		if (profile_id == profile) {
+			p11_message (_("profile already exists on the token"));
+			goto cleanup;
+		}
+	}
+
+	p11_kit_iter_free (iter);
+	iter = p11_kit_iter_new (uri, P11_KIT_ITER_WITH_MODULES);
+	if (iter == NULL) {
+		p11_message (_("failed to initialize iterator"));
+		goto cleanup;
+	}
+
+	p11_kit_iter_begin (iter, modules);
+	rv = p11_kit_iter_next (iter);
+	if (rv != CKR_OK) {
+		p11_message (_("failed to find the module"));
+		goto cleanup;
+	}
+
+	session = p11_kit_iter_get_session (iter);
+	module = p11_kit_iter_get_module (iter);
+
+	rv = module->C_CreateObject (session, object_template, 2, &object);
```

Comment:
nit: maybe it's a little more future proof to use `sizeof(object_template)/sizeof(object_template[0])` instead of 2?